lambda/translcomprehension.ml

```diff
@@ -2,304 +2,406 @@ open Lambda
-    | Array_of_elements -> arrs
-    | Array_of_arrays(arrs) -> arrs
-    | Array_of_filtered_arrays(arrs) -> arrs
+      With_size { body; raise_count }
+
+let sub_array ~loc src src_pos len =
+  let prim =
+    Primitive.simple ~name:"caml_array_sub" ~arity:3 ~alloc:true
   in
-  match res_kind with
-  | Pgenarray ->
-    Llet(Variable, Pgenval, res, empty_arr ~loc,
-      Llet(Variable, Pintval, counter, (int 0),
-        Lsequence(transl_for arr arr_len arrs, Lvar(res))))
-  | Pintarray | Pfloatarray | Paddrarray as kind ->
-    Llet(Strict, Pgenval, res, make_array_of_kind ~loc kind (Lvar(total_len)),
-      Llet(Variable, Pintval, counter, (int 0),
-        Lsequence(transl_for arr arr_len arrs, Lvar(res))))
+  Lprim (Pccall prim, [src; src_pos; len], loc)
 
-let transl_arr_comprehension body blocks ~array_kind ~scopes ~loc ~transl_exp =
-  (*The global counter is required when the size of the result array is not
-    known in advance.*)
-  let global_counter =
-    match blocks with
-    | [{ guard = None; _}] -> None
-    | _ -> Some (Ident.create_local "global_counter")
+let transl_single_arr_block ~transl_exp ~loc ~scopes
+      block body array_kind value_kind =
+  let body =
+    transl_arr_block ~transl_exp ~loc ~scopes None
+      (Without_size {body}) array_kind value_kind block
   in
-  let base_block =
-    transl_block ~transl_exp ~loc ~scopes global_counter
-      (Unguarded(transl_exp ~scopes body, array_kind),
-        Array_of_elements)
-      (List.hd blocks)
+  match body with
+  | Without_size { body } -> body
+  | With_size { body; raise_count } ->
+      let array_var = Ident.create_local "array" in
+      let len_var = Ident.create_local "len" in
+      Lstaticcatch(body,
+          (raise_count, [(array_var, Pgenval); (len_var, Pintval)]),
+          sub_array ~loc (Lvar array_var) (int 0) (Lvar len_var))
+
+type intermediate_array_shape =
+  | Array_of_elements
+  | Array_of_arrays of intermediate_array_shape
+  | Array_of_filtered_arrays of intermediate_array_shape
+
+let concat_arrays ~loc arr kind shape global_count_var =
+  let res_var = Ident.create_local "res" in
+  let res_binding =
+    make_array ~loc ~kind ~size:(Lvar global_count_var) ~array:res_var
   in
-  let translated_blocks, arrs =
-    List.fold_left
-      (fun acc el ->
-        transl_block  ~transl_exp  ~loc ~scopes None acc el )
-      base_block (List.tl blocks)
+  let counter_var = Ident.create_local "counter" in
+  let counter_binding = make_counter counter_var in
+  let rec loop shape arr_var len_var =
+    let kind =
+      match shape with
+      | Array_of_elements -> kind
+      | Array_of_arrays _ | Array_of_filtered_arrays _ -> Paddrarray
+    in
+    let len_var, bindings =
+      match len_var with
+      | Some var -> var, []
+      | None ->
+        let var = Ident.create_local "len" in
+        let init = Lprim((Parraylength kind), [Lvar(arr_var)], loc) in
+        let binding = binding Alias Pintval var init in
```

Comment:
This len_var is possibly used more than one time so it should be Strict. (in the Array_of_elements case)

Reply:
This is fine as the `init` expression is pure.